source/MaterialXRuntime/Private/Commands/PvtSetMetadataCmd.cpp

```diff
@@ -0,0 +1,151 @@
+//
+// TM & (c) 2020 Lucasfilm Entertainment Company Ltd. and Lucasfilm Ltd.
+// All rights reserved.  See LICENSE.txt for license.
+//
+
+#include <MaterialXRuntime/Private/Commands/PvtSetMetadataCmd.h>
+#include <MaterialXRuntime/Private/PvtCommand.h>
+
+#include <MaterialXRuntime/RtPrim.h>
+#include <MaterialXRuntime/RtTypeDef.h>
+#include <MaterialXRuntime/RtValue.h>
+
+namespace MaterialX
+{
+
+PvtSetMetadataCmd::PvtSetMetadataCmd(RtObject& obj, const RtToken& name, const RtValue& value)
+    : PvtCommand()
+    , _obj(obj)
+    , _name(name)
+    , _value(value)
+    , _oldValue()
+    , _metadataCreated(false)
+{
+}
+
+PvtCommandPtr PvtSetMetadataCmd::create(RtObject& obj, const RtToken& name, const RtValue& value)
+{
+    return std::make_shared<PvtSetMetadataCmd>(obj, name, value);
+}
+
+void PvtSetMetadataCmd::execute(RtCommandResult& result)
+{
+    if (_obj.isValid())
+    {
+        try
+        {
+            // Send message that the metadata is changing
+            msg().sendSetMetadataMessage(_obj, _name, _value);
+
+            // Do we need to create the metadata or does it already exist?
+            if (!_obj.getMetadata(_name))
```

Comment:
Here you should use the "typed" version of getMetadata (passing in the expected datatype as second argument). To detect the case where the metadata exists but has the wrong type. If this is not handled unexpected behavior can happen, for example trying to copy a string into an area allocated for storing just a float.